src/Passes/Source/RuleTransformationPass/Tests/Unit/LoopUnrolling.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Generators/DefaultProfileGenerator.hpp"
+#include "Rules/Factory.hpp"
+#include "TestTools/IrManipulationTestHelper.hpp"
+#include "gtest/gtest.h"
+
+#include "Llvm/Llvm.hpp"
+
+#include <functional>
+
+using namespace microsoft::quantum;
+
+namespace
+{
+using IrManipulationTestHelperPtr = std::shared_ptr<IrManipulationTestHelper>;
+IrManipulationTestHelperPtr newIrManip(std::string const& script)
+{
+    IrManipulationTestHelperPtr ir_manip = std::make_shared<IrManipulationTestHelper>();
+
+    ir_manip->declareOpaque("Qubit");
+    ir_manip->declareOpaque("Result");
+    ir_manip->declareOpaque("Array");
+    ir_manip->declareOpaque("Tuple");
+    ir_manip->declareOpaque("Range");
+    ir_manip->declareOpaque("Callable");
+    ir_manip->declareOpaque("String");
+
+    ir_manip->declareFunction("%Qubit* @__quantum__rt__qubit_allocate()");
+    ir_manip->declareFunction("void @__quantum__rt__qubit_release(%Qubit*)");
+    ir_manip->declareFunction("void @__quantum__qis__h__body(%Qubit*)");
+
+    ir_manip->declareFunction("i64 @TeleportChain__Calculate__body(i64, %Qubit*)");
+
+    if (!ir_manip->fromBodyString(script))
+    {
+        llvm::errs() << ir_manip->generateScript(script) << "\n\n";
+        llvm::errs() << ir_manip->getErrorMessage() << "\n";
+        exit(-1);
+    }
+    return ir_manip;
+}
+
+} // namespace
+
+// Single allocation with action and then release
+TEST(RuleTransformationPass, LoopUnroll)
+{
+    auto ir_manip = newIrManip(R"script(
```

Comment:
I would prefer to split out larger snippets to compare against into separate files rather than inline them in the test code. It gets messy quickly, and anywhere where we had not split things such as this split out into separate files I believe we have had a desire to revise it later.

Reply:
In principle I agree with you, although I do not see it as a huge issue. Since we are using a combination of CMake and Google Test, I do not think that there is an elegant solution to this, but I am open to suggestions. Keep in mind that you would need the code snippet together with the expected instruction and unexpected instructions to be separated out.

How do you envision this will be structured? Where would these files live? And how would the program know where to find these files?

Reply:
I don't know about elegant, but the simple setup that has worked well is simply by having these files checked in in a subfolder and the tests have hardcoded paths to those files. Within dotnet, the files are considered to be content files that are part of the project and hence automatically copied to the build output folder. I don't know if there is a similar setup that can conveniently be used here.

Reply:
I am happy to have that discussion. I do think it is out of scope for the present PR as this would require a large change in how we deal with tests atm.

Reply:
I agree to defer this conversation to a separate PR, given that we have enough dependencies between PRs as it is.